twinify/napsu_mq/napsu_mq.py

```diff
@@ -0,0 +1,195 @@
+                                   cliques_to_include=column_feature_set)
+        queries = FullMarginalQuerySet(query_sets, dataframe.values_by_col)
+        queries = queries.get_canonical_queries()
+        mnjax = MarkovNetworkJax(dataframe.values_by_col, queries)
+        suff_stat = np.sum(queries.flatten()(dataframe.int_array), axis=0)
+
+        # determine Gaussian mech DP noise level for given privacy level
+        sensitivity = np.sqrt(2 * len(query_sets))
+        sigma_DP = privacy_accounting.sigma(epsilon, delta, sensitivity)
+
+        # add DP noise (according to Gaussian mechanism)
+        inference_rng, dp_rng = d3p.random.split(rng, 2)
+        dp_noise = d3p.random.normal(dp_rng, suff_stat.shape) * sigma_DP
+        dp_suff_stat = suff_stat + dp_noise
+
+        inference_rng = d3p.random.convert_to_jax_rng_key(inference_rng)
+
+        if use_laplace_approximation is True:
+            approx_rng, mcmc_rng = jax.random.split(inference_rng, 2)
+            laplace_approx, success = mei.run_numpyro_laplace_approximation(approx_rng, dp_suff_stat, n, sigma_DP, mnjax)
+            mcmc, backtransform = mei.run_numpyro_mcmc_normalised(
+                mcmc_rng, dp_suff_stat, n, sigma_DP, mnjax, laplace_approx, num_samples=2000, num_warmup=800, num_chains=4
+            )
+            inf_data = az.from_numpyro(mcmc, log_likelihood=False)
+            posterior_values = inf_data.posterior.stack(draws=("chain", "draw"))
+            posterior_values = backtransform(posterior_values.norm_lambdas.values.transpose())
+
+        else:
+            mcmc = mei.run_numpyro_mcmc(
+                inference_rng, dp_suff_stat, n, sigma_DP, mnjax, num_samples=2000, num_warmup=800, num_chains=4
+            )
+            inf_data = az.from_numpyro(mcmc, log_likelihood=False)
+            posterior_values = inf_data.posterior.stack(draws=("chain", "draw"))
+            posterior_values = posterior_values.lambdas.values.transpose()
+
+        return NapsuMQResult(mnjax, posterior_values, category_mapping)
+
+
+class NapsuMQResult(InferenceResult):
+
+    def __init__(self, markov_network: MarkovNetworkJax, posterior_values: jnp.ndarray,
+                 category_mapping: Mapping) -> None:
+        super().__init__()
+        self._markov_network = markov_network
+        self._posterior_values = posterior_values
+        self._category_mapping = category_mapping
+
+    @property
+    def markov_network(self) -> MarkovNetworkJax:
+        return self._markov_network
+
+    @property
+    def posterior_values(self) -> jnp.ndarray:
+        return self._posterior_values
+
+    @property
+    def category_mapping(self) -> Mapping:
+        return self._category_mapping
+
+    def _store_to_io(self, write_io: BinaryIO) -> None:
+        assert write_io.writable()
+        result = dill.dumps(self)
+        return NapsuMQResultIO.store_to_io(write_io, result)
+
+    @classmethod
+    def _load_from_io(cls, read_io: BinaryIO) -> 'NapsuMQResult':
```

Comment:
due to recent change to base class: Please add `**kwargs`  to argument list (even if the method ignores them)